tests/test_freshdesk_bookmarks.py

```diff
@@ -0,0 +1,218 @@
+        a string formatted utc datetime,
+        in order to compare aginast json formatted datetime values
+        """
+        date_object = dateutil.parser.parse(date_str)
+        date_object_utc = date_object.astimezone(tz=pytz.UTC)
+        return datetime.strftime(date_object_utc, "%Y-%m-%dT%H:%M:%SZ")
+
+    def calculated_states_by_stream(self, current_state):
+        """
+        Look at the bookmarks from a previous sync and set a new bookmark
+        value based off timedelta expectations. This ensures the subsequent sync will replicate
+        at least 1 record but, fewer records than the previous sync.
+
+        Sufficient test data is required for this test to cover a given stream.
+        An incremental replication stream must have at least two records with
+        replication keys that differ by some time span.
+
+        If the test data is changed in the future this may break expectations for this test.
+        """
+        timedelta_by_stream = {stream: [2,0,0]  # {stream_name: [days, hours, minutes], ...}
+                               for stream in self.expected_streams()}
+        # Works with static start date to go back to 2017 Feb 8 22:xx:xx and pick up 5 of 6 records
+        # including 2 of 3 conversations assocated with those 5 ticket records
+        timedelta_by_stream['tickets'] = [698, 17, 26]
+        timedelta_by_stream['companies'] = [0, 0, 10]
+        timedelta_by_stream['agents'] = [0, 1, 0]
+
+        #timedelta_by_stream['conversations'] = [1, 0 , 0]  # Child stream of tickets, no bookmarks
+        # TODO Add conversations,time_entries, satisfaction_ratings streams
+
+        # BUG https://jira.talendforge.org/browse/TDL-17559.  Redefining state to be closer to
+        # expected format so the underlying code wont have to change as much after the JIRA fix
+        current_state = {'bookmarks': current_state}
+        del current_state['bookmarks']['tickets_deleted']  # Delete unexpected streams
+        del current_state['bookmarks']['tickets_spam']     # generated by filter?
+
+        # Keep existing format for this method so it will work after bug fix
+        stream_to_calculated_state = {stream: "" for stream in current_state['bookmarks'].keys()}
+        for stream, state_value in current_state['bookmarks'].items():
+            state_as_datetime = dateutil.parser.parse(state_value)
+
+            days, hours, minutes = timedelta_by_stream[stream]
+            calculated_state_as_datetime = state_as_datetime - timedelta(days=days, hours=hours, minutes=minutes)
+
+            state_format = self.BOOKMARK_FORMAT
+            calculated_state_formatted = datetime.strftime(calculated_state_as_datetime, state_format)
+
+            stream_to_calculated_state[stream] = calculated_state_formatted
+
+        return stream_to_calculated_state
+
+    # function for verifying the date format
+    def is_expected_date_format(self, date):
```

Comment:
I don't think this method is used anywhere, can be deleted

Reply:
Deleted